service/matching/taskQueueManager.go

```diff
@@ -401,12 +405,28 @@ func (c *taskQueueManagerImpl) AddTask(
 		return false, errRemoteSyncMatchFailed
 	}
 
-	_, err = c.taskWriter.appendTask(params.execution, taskInfo)
+	// Ensure that tasks with the "default" versioning directive get spooled in the unversioned queue as they not
+	// associated with any version set until their execution is touched by a version specific worker.
+	// "compatible" tasks OTOH are associated with a specific version set and should be stored along with all tasks for
+	// that version set.
+	// The task queue default set is dynamic and applies only at dispatch time. Putting "default" tasks into version set
+	// specific queues could cause them to get stuck behind "compatible" tasks when they should be able to progress
+	// independently.
+	if _, ok := taskInfo.VersionDirective.GetValue().(*taskqueuespb.TaskVersionDirective_UseDefault); ok {
+		err = params.unversionedTqm.SpoolTask(params)
```

Comment:
do you think this might be cleaner if we moved the fallback into matchingEngine? I guess there are a bunch of call sites so you'd have to factor it out...

it's okay for now, I can clean it up later (I'm already doing some related cleanups in upcoming PRs)

Reply:
Agree it would be cleaner but I'd rather avoid a big last minute change.